### PR TITLE
Visitor api ti serialization

### PR DESCRIPTION
**Pull request by sdurawa**
TensorIterator serialization. 
Ticket: 45050

### CODE REVIEW

File: inference-engine/src/readers/ir_reader/ie_ir_parser.cpp [hunk 1]
```diff
@@ -984,7 +984,6 @@ template <>
 std::shared_ptr<ngraph::Node> V10Parser::LayerCreator<ngraph::op::v0::LSTMCell>::createLayer(
     const ngraph::OutputVector& inputs, const pugi::xml_node& node, const Blob::CPtr& weights,
     const GenericLayerParams& layerParsePrms) {
-    checkParameters(inputs, layerParsePrms, 6);
```
Comment (ilyachur) [inline inference-engine/src/readers/ir_reader/ie_ir_parser.cpp:987]: Why did you remove this check?

File: inference-engine/src/transformations/src/transformations/serialize.cpp [hunk 2]
```diff
@@ -148,6 +238,19 @@ class XmlSerializer : public ngraph::AttributeVisitor {
         m_xml_node.append_attribute(name.c_str())
             .set_value(create_atribute_list(adapter).c_str());
     }
+    void on_adapter(
+        const std::string& name,
+        ngraph::ValueAccessor<std::shared_ptr<Function>>& adapter) override {
+        pugi::xml_document xml_body;
```
Comment (ilyachur) [inline inference-engine/src/transformations/src/transformations/serialize.cpp:244]: Can we use this `on_adapter()` in order to serialize all functions?

File: inference-engine/src/readers/ir_reader/ie_ir_parser.cpp (hunk 1 above)
Comment (sdurawa) [reply inference-engine/src/readers/ir_reader/ie_ir_parser.cpp:987]: MO provides LSTMCell with 6 inputs. Inside the LSTMCell op we are adding 7th input (peephole). After serialization LSTMCell IR format consists of 7 inputs and test is failing due to this check. Hardocded value in check is not correct. Besides this will be removed when LSTMCell will be deserialized.

File: inference-engine/src/transformations/src/transformations/serialize.cpp (hunk 2 above)
Comment (sdurawa) [reply inference-engine/src/transformations/src/transformations/serialize.cpp:244]: It is already used in order to serialize nGraph function to fill body part for tensor iterator.

Comment (ilyachur) [reply inference-engine/src/transformations/src/transformations/serialize.cpp:244]: I believe that we can use it inside the Serialization pass.

File: inference-engine/src/readers/ir_reader/ie_ir_parser.cpp (hunk 1 above)
Comment (ilyachur) [reply inference-engine/src/readers/ir_reader/ie_ir_parser.cpp:987]: We shouldn't remove this check, because Serializer and Deserializer should be align with OpenVINO spec.
In OpenVINO spec [`LSTMCell`](https://docs.openvinotoolkit.org/latest/openvino_docs_ops_sequence_LSTMCell_1.html) has only 6 inputs.
@itikhono @lazarevevgeny Do you remember why we have this difference between specification and the nGraph operation?

File: inference-engine/src/transformations/src/transformations/serialize.cpp [hunk 3]
```diff
@@ -423,7 +530,7 @@ void ngfunction_2_irv10(pugi::xml_document& doc,
 
         const auto data_attr_size =
```
Comment (pelszkow) [inline inference-engine/src/transformations/src/transformations/serialize.cpp:531]: @jdanieck 
Can we switch to:
```
const bool no_attributes = data.attributes().begin() == data.attributes().end();
```
This will skip unnecessary traverse whole list of attributes.

File: inference-engine/src/readers/ir_reader/ie_ir_parser.cpp (hunk 1 above)
Comment (lazarevevgeny) [reply inference-engine/src/readers/ir_reader/ie_ir_parser.cpp:987]: @itikhono , could you comment on that? I don't remember the whole story.

File: inference-engine/src/transformations/src/transformations/serialize.cpp (hunk 3 above)
Comment (jdanieck) [reply inference-engine/src/transformations/src/transformations/serialize.cpp:531]: @pelszkow intent of this code is to not introduce <data> node to IR if there are no attributes, so if std::distance doesn't do the same under the hood we can switch to the code you proposed.

@sdurawa why are we skipping this check for TensorIterator?

File: inference-engine/src/readers/ir_reader/ie_ir_parser.cpp (hunk 1 above)
Comment (itikhono) [reply inference-engine/src/readers/ir_reader/ie_ir_parser.cpp:987]: LSTMCell v0 version has 7 inputs (one of these is the default input for peepholes, it is not used actually, there is a constructor of v0 version with 6 inputs to match the specification, default peepholes input adds in the definition of this constructor ), it's a mistake and it was fixed in LSTMCell v4 version in 2021.1 Release. 
I suppose there shouldn't be any problems with v4 version. The issue is relevant only for v0 version, I guess we need to have some workaround for that in serialization code (handle a special case for LSTMCell v0), maybe we should always serialize only 6 inputs for this version. @ilyachur @lazarevevgeny what do you think?

File: inference-engine/tests/functional/inference_engine/ir_serialization/serialize.cpp [hunk 4]
```diff
@@ -48,18 +73,32 @@ TEST_P(SerializationTest, CompareFunctions) {
...
+                        std::make_tuple("experimental_detectron_detection_output.xml", 0, nullptr),
+                        std::make_tuple("nms5.xml", 0, nullptr),
+                        std::make_tuple("ti_negative_stride.xml", 3149864, +[](InferenceEngine::Blob::Ptr& weights) {
+                                                        auto *data = weights->buffer().as<int64_t *>();
+                                                        data[0] = 1;
+                                                        data[1] = 512;
+                                                        data[393730] = 1;
+                                                        data[393731] = 1;
+                                                        data[393732] = 256;}),
+                        std::make_tuple("ti_resnet.xml", 8396840, +[](InferenceEngine::Blob::Ptr& weights) {
```
Comment (jdanieck) [inline inference-engine/tests/functional/inference_engine/ir_serialization/serialize.cpp:91]: @sdurawa I think it's better to move TensorIterator tests to separate file and revert tests to previous state*. For now it seems that dynamically generated weights are needed for TI & Loop only, so this is a special case and it's better to not complicate test logic for general case. 

** I've noticed an issue in previous code. CNNNetwork is constructed from xml file only, but should also take bin file into account. Please try to fix this. There are .bin files for most of the models tested here (e.g. add_abc.bin) if it is not present in repo then use empty string.  Something like this:
 ```
testing::Values(std::make_tuple("add_abc.xml","add_abc.bin"),  std::make_tuple("nms5.xml", ""))
```

File: inference-engine/tests/functional/shared_test_classes/include/shared_test_classes/base/layer_test_utils.hpp [hunk 5]
```diff
@@ -31,6 +31,8 @@
 
 namespace LayerTestsUtils {
 
+constexpr std::size_t maxFileNameLength = 140;
```
Comment (jdanieck) [inline inference-engine/tests/functional/shared_test_classes/include/shared_test_classes/base/layer_test_utils.hpp:34]: Please add a comment with explanation why this limit is needed and why it is set 140.

File: inference-engine/src/transformations/src/transformations/serialize.cpp (hunk 3 above)
Comment (sdurawa) [reply inference-engine/src/transformations/src/transformations/serialize.cpp:531]: TensorIterator does not have attributes in "data" section. For TI "data" layer (passed to XmlSerializer) is renamed to "body" layer and filled with nGraph function. This is the reason of not removing an empty layer for TI.

File: inference-engine/tests/functional/inference_engine/ir_serialization/serialize.cpp (hunk 4 above)
Comment (sdurawa) [reply inference-engine/tests/functional/inference_engine/ir_serialization/serialize.cpp:91]: @jdanieck, current serialization test class takes into account binary file without specific inclusion in make_tuple(). In setUp() method: 
`m_out_xml_path = test_name + ".xml";`
`m_out_bin_path = test_name + ".bin";`
When performing a test following code is used:
`expected = ie.ReadNetwork(m_model_path);`
It reads bin file with the same name as xml and if bin file with the same name was not found, will load IR without weights.

Comment (jdanieck) [reply inference-engine/tests/functional/inference_engine/ir_serialization/serialize.cpp:91]: Ok then, thx Szymon. Didn't notice that ReadNetwork will create this path internally. So knowing that I'd say that calling ReadNetwork with bin path explicitly is not mandatory, but either way it is preferred in my mind because it shows intent better and you don't need to know API details to understand behaviour on the caller side. Your call, I won't insist on this change.

File: inference-engine/src/transformations/src/transformations/serialize.cpp [hunk 6]
```diff
@@ -69,16 +75,106 @@ class XmlSerializer : public ngraph::AttributeVisitor {
...
+                  std::string& node_type_name,
+                  const std::map<std::string, ngraph::OpSet>& custom_opsets)
         : m_xml_node(data)
         , m_bin_data(bin_data)
-        , m_node_type_name(node_type_name) {
+        , m_node_type_name(node_type_name)
+        , m_custom_opsets(custom_opsets) {
+    }
+
+    void map_type_from_body(const pugi::xml_node& xml_node,
```
Comment (pelszkow) [inline inference-engine/src/transformations/src/transformations/serialize.cpp:86]: Is any reason why `output` is taking by reference? For me is cleaner when function return by value.

File: inference-engine/src/transformations/src/transformations/serialize.cpp [hunk 7]
```diff
@@ -69,16 +75,106 @@ class XmlSerializer : public ngraph::AttributeVisitor {
...
+        const std::string map_type, std::vector<std::string>& output) {
```
Comment (pelszkow) [inline inference-engine/src/transformations/src/transformations/serialize.cpp:87]: I thing reference is missing here (`const std::string& map_type`).

File: inference-engine/src/transformations/src/transformations/serialize.cpp [hunk 8]
```diff
@@ -69,16 +75,106 @@ class XmlSerializer : public ngraph::AttributeVisitor {
...
+        if (!map_type.compare("Result") || !map_type.compare("Parameter")) {
```
Comment (pelszkow) [inline inference-engine/src/transformations/src/transformations/serialize.cpp:88]: IMHO, `compare` looks odd. 
-> https://en.cppreference.com/w/cpp/string/basic_string/operator_cmp -> `9` version takes `const char*`

File: inference-engine/src/transformations/src/transformations/serialize.cpp [hunk 9]
```diff
@@ -69,16 +75,106 @@ class XmlSerializer : public ngraph::AttributeVisitor {
...
+            for (pugi::xml_node node : xml_node.child("body").child("layers")) {
+                auto param = std::strcmp(node.attribute("type").value(), map_type.c_str());
```
Comment (pelszkow) [inline inference-engine/src/transformations/src/transformations/serialize.cpp:90]: If `map_type` is `std::string` now string comparator might be used here.

File: inference-engine/src/transformations/src/transformations/serialize.cpp [hunk 10]
```diff
@@ -148,6 +244,20 @@ class XmlSerializer : public ngraph::AttributeVisitor {
...
             .set_value(create_atribute_list(adapter).c_str());
     }
+    void on_adapter(
+        const std::string& name,
+        ngraph::ValueAccessor<std::shared_ptr<Function>>& adapter) override {
+        pugi::xml_document xml_body;
+
+        ngfunction_2_irv10(xml_body, m_bin_data, *adapter.get(), m_custom_opsets);
+
+        if (!name.compare("body")) {
```
Comment (pelszkow) [inline inference-engine/src/transformations/src/transformations/serialize.cpp:254]: Consider `operator ==` ?

File: inference-engine/tests/functional/inference_engine/ir_serialization/tensor_iterator.cpp [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+        std::remove(m_out_xml_path.c_str());
+        std::remove(m_out_xml_path.c_str());
+    }
+
+    void serialize_and_compare(const std::string& model_path, InferenceEngine::Blob::Ptr weights) {
+        std::stringstream buffer;
+        InferenceEngine::Core ie;
+
+        std::ifstream model(model_path);
+        buffer << model.rdbuf();
```
Comment (pelszkow) [inline inference-engine/tests/functional/inference_engine/ir_serialization/tensor_iterator.cpp:34]: Can be add `ASSERT_TRUE(model);` here for test the test?

File: inference-engine/src/readers/ir_reader/ie_ir_parser.cpp (hunk 1 above)
Comment (lazarevevgeny) [reply inference-engine/src/readers/ir_reader/ie_ir_parser.cpp]: Probably serializer could dump the v0 operation in a correct format with 6 inputs. Is it possible? Anyway, any working solution for the buggy v0 operation should be fine.

Comment (sdurawa) [reply inference-engine/src/readers/ir_reader/ie_ir_parser.cpp]: WA on Serializer side provided.

File: inference-engine/src/transformations/src/transformations/serialize.cpp (hunk 2 above)
Comment (sdurawa) [reply inference-engine/src/transformations/src/transformations/serialize.cpp:250]: on_adapter() is implemented in Serialization pass.

File: inference-engine/tests/functional/shared_test_classes/include/shared_test_classes/base/layer_test_utils.hpp (hunk 5 above)
Comment (sdurawa) [reply inference-engine/tests/functional/shared_test_classes/include/shared_test_classes/base/layer_test_utils.hpp:35]: Commented.

File: inference-engine/src/transformations/src/transformations/serialize.cpp (hunk 3 above)
Comment (sdurawa) [reply inference-engine/src/transformations/src/transformations/serialize.cpp]: Skipping for TI is removed.

File: inference-engine/tests/functional/inference_engine/ir_serialization/serialize.cpp (hunk 4 above)
Comment (sdurawa) [reply inference-engine/tests/functional/inference_engine/ir_serialization/serialize.cpp]: New approach is implemented.

File: inference-engine/src/transformations/src/transformations/serialize.cpp (hunk 6 above)
Comment (sdurawa) [reply inference-engine/src/transformations/src/transformations/serialize.cpp]: force of habit, I have changed implementation and also added check for empty vectors.

File: inference-engine/src/transformations/src/transformations/serialize.cpp (hunk 8 above)
Comment (sdurawa) [reply inference-engine/src/transformations/src/transformations/serialize.cpp]: right, those comparisons do not look like naturally. Switched to ==.

File: inference-engine/src/transformations/src/transformations/serialize.cpp (hunk 9 above)
Comment (sdurawa) [reply inference-engine/src/transformations/src/transformations/serialize.cpp]: Corrected.

File: inference-engine/src/transformations/src/transformations/serialize.cpp (hunk 10 above)
Comment (sdurawa) [reply inference-engine/src/transformations/src/transformations/serialize.cpp]: Corrected.

File: inference-engine/src/transformations/src/transformations/serialize.cpp [hunk 12]
```diff
@@ -148,6 +250,27 @@ class XmlSerializer : public ngraph::AttributeVisitor {
...
+        if (name == "body") {
+            xml_body.first_child().remove_attribute("name");
+            xml_body.first_child().remove_attribute("version");
+            xml_body.first_child().set_name(name.c_str());
+            // TI, Loop do not have attributtes as regular ops, it is necessary to append "body"
+            // to layer above (m_xml_node.parent()) as in ngfunction_2_irv10() layer (m_xml_node) with empty attributes
+            // is removed.
+            m_xml_node.parent().append_copy(xml_body.first_child());
+        } else if (name == "net") {
+            m_xml_node.append_copy(xml_body.first_child());
```
Comment (ilyachur) [inline inference-engine/src/transformations/src/transformations/serialize.cpp:269]: Why do we need  to call this `append_copy()`?

File: inference-engine/src/transformations/src/transformations/serialize.cpp [hunk 13]
```diff
@@ -435,6 +559,12 @@ void ngfunction_2_irv10(pugi::xml_document& doc,
                 NGRAPH_CHECK(i.get_partial_shape().is_static(),
                              "Unsupported dynamic input shape in ", node);
 
+                // WA for LSTMCellv0, peephole input shall not be serialized
+                if (node_type_name == "LSTMCell" && (i.get_index() == 6)) {
```
Comment (ilyachur) [inline inference-engine/src/transformations/src/transformations/serialize.cpp:563]: I think that we need to do this thing only for LSTMCell v0, I propose to check the version of the operation too. Or we can use `NodeTypeInfo` to check type and version.

File: inference-engine/src/transformations/src/transformations/serialize.cpp [hunk 14]
```diff
@@ -406,24 +529,25 @@ void ngfunction_2_irv10(pugi::xml_document& doc,
             layer.append_attribute("version").set_value(
                 get_opset_name(node, custom_opsets).c_str());
         }
+
         // <layers/data>
         pugi::xml_node data = layer.append_child("data");
+        std::string node_type_name{node->get_type_name()};
```
Comment (ilyachur) [inline inference-engine/src/transformations/src/transformations/serialize.cpp:535]: Do we change the `node_type_name` or it can be constant?

File: inference-engine/src/transformations/src/transformations/serialize.cpp [hunk 15]
```diff
@@ -490,7 +630,11 @@ bool pass::Serialize::run_on_function(std::shared_ptr<ngraph::Function> f) {
     NGRAPH_CHECK(bin_file, "Can't open bin file: \"" + m_binPath + "\"");
     switch (m_version) {
     case Version::IR_V10:
-        ngfunction_2_irv10(xml_doc, bin_file, *f, m_custom_opsets);
+        {
+            std::string name = "net";
```
Comment (ilyachur) [inline inference-engine/src/transformations/src/transformations/serialize.cpp:634]: nitpick: I think we use pass "net" in-place (without local variables).

File: inference-engine/tests/functional/inference_engine/ir_serialization/serialize.cpp [hunk 16]
```diff
@@ -48,18 +58,18 @@ TEST_P(SerializationTest, CompareFunctions) {
...
-                        std::make_tuple("nms5.xml"),
-                        std::make_tuple("shape_of.xml")));
+        testing::Values(std::make_tuple("add_abc.xml", "add_abc.bin"),
+                        std::make_tuple("add_abc_f64.xml", ""),
+                        std::make_tuple("split_equal_parts_2d.xml", "split_equal_parts_2d.bin"),
+                        std::make_tuple("addmul_abc.xml", "addmul_abc.bin"),
+                        std::make_tuple("add_abc_initializers.xml", "add_abc_initializers.bin"),
+                        std::make_tuple("experimental_detectron_roi_feature_extractor.xml", ""),
+                        std::make_tuple("experimental_detectron_detection_output.xml", ""),
+                        std::make_tuple("nms5.xml", "nms5.bin"),
```
Comment (ilyachur) [inline inference-engine/tests/functional/inference_engine/ir_serialization/serialize.cpp:68]: Why do we need to specialize the bin files?

File: inference-engine/src/transformations/src/transformations/serialize.cpp (hunk 12 above)
Comment (sdurawa) [reply inference-engine/src/transformations/src/transformations/serialize.cpp:269]: It copies xml_body (which contains nGraph function) into empty XmlSerializer::m_xml_node.

File: inference-engine/src/transformations/src/transformations/serialize.cpp (hunk 14 above)
Comment (sdurawa) [reply inference-engine/src/transformations/src/transformations/serialize.cpp:535]: it is changed in `visit_exec_graph_node()`

File: inference-engine/src/transformations/src/transformations/serialize.cpp (hunk 15 above)
Comment (sdurawa) [reply inference-engine/src/transformations/src/transformations/serialize.cpp:634]: it is problematic for the `XmlSerializer` as its 3rd input `std::string& m_node_type_name` is not const (due to WA for GenericIE op) and "net" name cannot be used  in-place.

File: inference-engine/tests/functional/inference_engine/ir_serialization/serialize.cpp (hunk 16 above)
Comment (sdurawa) [reply inference-engine/tests/functional/inference_engine/ir_serialization/serialize.cpp:68]: It is a matter of usage, this way better explains idea of serialization test without exploring how it works inside, behavior is the same.

File: inference-engine/src/transformations/src/transformations/serialize.cpp (hunk 12 above)
Comment (sdurawa) [reply inference-engine/src/transformations/src/transformations/serialize.cpp]: ngfunction_2_irv10 API is changed, hence no more copy is required.

File: inference-engine/src/transformations/src/transformations/serialize.cpp (hunk 13 above)
Comment (sdurawa) [reply inference-engine/src/transformations/src/transformations/serialize.cpp]: Corrected.

File: inference-engine/tests/functional/inference_engine/ir_serialization/tensor_iterator.cpp (hunk 11 above)
Comment (sdurawa) [reply inference-engine/tests/functional/inference_engine/ir_serialization/tensor_iterator.cpp:35]: Assert is added.

File: inference-engine/src/transformations/src/transformations/serialize.cpp [hunk 17]
```diff
@@ -85,16 +91,112 @@ class XmlSerializer : public ngraph::AttributeVisitor {
...
+        , m_custom_opsets(custom_opsets) {
+    }
+
+    std::vector<std::string> map_type_from_body(const pugi::xml_node& xml_node,
+        const std::string& map_type) {
+        std::vector<std::string> output;
+        if (map_type == "Result" || map_type == "Parameter") {
+            for (pugi::xml_node node : xml_node.child("body").child("layers")) {
+                auto param = map_type.compare(node.attribute("type").value());
+                if (!param) {
```
Comment (ilyachur) [inline inference-engine/src/transformations/src/transformations/serialize.cpp:108]: nitpick: Looks like we can do a compare in-place.

File: inference-engine/src/transformations/src/transformations/serialize.cpp [hunk 18]
```diff
@@ -85,16 +91,112 @@ class XmlSerializer : public ngraph::AttributeVisitor {
...
+
+        return output;
     }
 
     void on_adapter(const std::string& name,
                     ngraph::ValueAccessor<void>& adapter) override {
         (void)name;
-        (void)adapter;
+
+        if (m_xml_node.parent().child("body")) {
```
Comment (jdanieck) [inline inference-engine/src/transformations/src/transformations/serialize.cpp:124]: This comment is about potential refactoring - I don't expect to do it in this PR but we could take a note to do this after release when we'll have more time.

This if is too complicated and because of it it's hard to understand.  I suggest to extract 2 separate helper functions and have something like this:

```
on_adapter(const std::string& name, ngraph::ValueAccessor<void>& adapter)
{
    if (m_xml_node.parent().child("body"))
    {
        if (name == "input_descriptions") 
            input_descriptions_on_adapter();
        else if (name == "output_descriptions")  
            output_descriptions_on_adapter();                
    }
}
```

**Review by ilyachur — APPROVED** [review]

**Review by jdanieck — APPROVED** [review]

File: inference-engine/src/transformations/src/transformations/serialize.cpp [hunk 19]
```diff
@@ -452,6 +571,15 @@ void ngfunction_2_irv10(pugi::xml_document& doc,
                 NGRAPH_CHECK(i.get_partial_shape().is_static(),
                              "Unsupported dynamic input shape in ", node);
 
+                // WA for LSTMCellv0, peephole input shall not be serialized
+                if (i.get_index() == 6) {
+                    auto type_info = node->get_type_info();
+                    if (!strcmp(type_info.name, "LSTMCell") && type_info.version == 0) {
+                        port_id++;
```
Comment (itikhono) [inline inference-engine/src/transformations/src/transformations/serialize.cpp:578]: Not sure that we need to increment port_id and not just skip.  Could you please explain?

File: inference-engine/src/transformations/src/transformations/serialize.cpp (hunk 17 above)
Comment (sdurawa) [reply inference-engine/src/transformations/src/transformations/serialize.cpp]: Corrected.

File: inference-engine/src/transformations/src/transformations/serialize.cpp [hunk 20]
```diff
@@ -85,16 +91,112 @@ class XmlSerializer : public ngraph::AttributeVisitor {
...
         , m_bin_data(bin_data)
-        , m_node_type_name(node_type_name) {
+        , m_node_type_name(node_type_name)
+        , m_custom_opsets(custom_opsets) {
+    }
+
+    std::vector<std::string> map_type_from_body(const pugi::xml_node& xml_node,
+        const std::string& map_type) {
+        std::vector<std::string> output;
+        if (map_type == "Result" || map_type == "Parameter") {
```
[the remark this answers is not in the document]
Comment (sdurawa) [reply inference-engine/src/transformations/src/transformations/serialize.cpp]: Corrected.

File: inference-engine/src/transformations/src/transformations/serialize.cpp (hunk 17 above)
[the remark this answers is not in the document]
Comment (sdurawa) [reply inference-engine/src/transformations/src/transformations/serialize.cpp]: Corrected.

File: inference-engine/src/transformations/src/transformations/serialize.cpp [hunk 21]
```diff
@@ -85,16 +91,112 @@ class XmlSerializer : public ngraph::AttributeVisitor {
...
+            // parameters and results from body are required for port_map attributes serialization
+            std::vector<std::string> parameter_mapping = map_type_from_body(m_xml_node.parent(), "Parameter");
+            std::vector<std::string> result_mapping = map_type_from_body(m_xml_node.parent(), "Result");
+
+            NGRAPH_CHECK(!parameter_mapping.empty() || !result_mapping.empty(), "No parameters or results found in body Function.");
+
+            // TI, Loop do not have attributtes as regular ops, it is necessary to append "port_map" and
+            // "back_edges" to layer above (m_xml_node.parent()) as in ngfunction_2_irv10() layer (here "m_xml_node")
+            // with empty attributes is removed.
+            if (auto a = ngraph::as_type<ngraph::AttributeAdapter<std::vector<std::shared_ptr
```
[the remark this answers is not in the document]
Comment (sdurawa) [reply inference-engine/src/transformations/src/transformations/serialize.cpp]: Corrected.

File: inference-engine/src/transformations/src/transformations/serialize.cpp [hunk 22]
```diff
@@ -85,16 +91,112 @@ class XmlSerializer : public ngraph::AttributeVisitor {
...
+                        <ngraph::op::util::SubGraphOp::InputDescription>>>>(&adapter)) {
+                pugi::xml_node port_map = m_xml_node.parent().find_child([](pugi::xml_node node) {return strcmp(node.name(), "port_map") == 0;});
```
[the remark this answers is not in the document]
Comment (sdurawa) [reply inference-engine/src/transformations/src/transformations/serialize.cpp]: Right, it is better solution.

File: inference-engine/src/transformations/src/transformations/serialize.cpp (hunk 19 above)
Comment (sdurawa) [reply inference-engine/src/transformations/src/transformations/serialize.cpp:575]: If port_id incrementation is skipped then we need to align back_edges with newer port numeration. It will complicate WA implementation for LSTMCell.

File: inference-engine/src/transformations/src/transformations/serialize.cpp (hunk 18 above)
Comment (sdurawa) [reply inference-engine/src/transformations/src/transformations/serialize.cpp:121]: I have created a jira ticket (46975) for all future improvements to not forget about it. This issue is already addressed as a subtask (46976).

File: inference-engine/src/transformations/src/transformations/serialize.cpp [hunk 23]
```diff
@@ -85,16 +91,112 @@ class XmlSerializer : public ngraph::AttributeVisitor {
...
+                    pugi::xml_node input = port_map.append_child("input");
+                    input.append_attribute("external_port_id").set_value(input_description->m_input_index);
+                    input.append_attribute("internal_layer_id").set_value(parameter_mapping[input_description->m_body_parameter_index].c_str());
+
+                    if (auto slice_input = as_type_ptr<ngraph::op::util::SubGraphOp::SliceInputDescription>(input_description)) {
+                        input.prepend_attribute("axis").set_value(slice_input->m_axis);
+                        if (slice_input->m_start) {
+                            input.append_attribute("start").set_value(slice_input->m_start);
+                        }
+                        if (slice_input->m_end != -1) {
```
[the remark this answers is not in the document]
Comment (sdurawa) [reply inference-engine/src/transformations/src/transformations/serialize.cpp:148]: Maybe we shall add this to ticket 46975 to not block releasing this feature too long before code freeze?

[the remark this answers is not in the document]
Comment (itikhono) [reply inference-engine/src/transformations/src/transformations/serialize.cpp:148]: Ok, please add this

File: inference-engine/src/transformations/src/transformations/serialize.cpp [hunk 24]
```diff
@@ -85,16 +91,109 @@ class XmlSerializer : public ngraph::AttributeVisitor {
...
+                            back_edges = m_xml_node.parent().insert_child_after("back_edges", port_map);
+                        }
+                        pugi::xml_node edge = back_edges.append_child("edge");
+                        edge.append_attribute("from-layer").set_value(result_mapping[merged_input->m_body_value_index].c_str());
+                        edge.append_attribute("to-layer").set_value(parameter_mapping[merged_input->m_body_parameter_index].c_str());
+                    }
+                }
+            } else if (const auto& a = ngraph::as_type<ngraph::AttributeAdapter<std::vector<std::shared_ptr
+                        <ngraph::op::util::SubGraphOp::OutputDescription>>>>(&adapter)) {
+                pugi::xml_node port_map = m_xml_node.parent().find_child([](pugi::xml_node node) {return strcmp(node.name(), "port_map") == 0;});
```
Comment (itikhono) [inline inference-engine/src/transformations/src/transformations/serialize.cpp:169]: .parent().child("port_map");

**Review by itikhono — APPROVED** [review]

File: inference-engine/src/transformations/src/transformations/serialize.cpp (hunk 23 above)
[the remark this answers is not in the document]
Comment (sdurawa) [reply inference-engine/src/transformations/src/transformations/serialize.cpp:148]: I have added this feature to ticket 46976 which already is addressing on_adapter() refactor.

File: inference-engine/src/transformations/src/transformations/serialize.cpp (hunk 24 above)
Comment (sdurawa) [reply inference-engine/src/transformations/src/transformations/serialize.cpp:169]: will be removed in Loop serialization as it will modify on_adapter() for input/output description.
